### PR TITLE
UCP/RNDV:Add put_zcopy in UCP RNDV protocol.

### PR DESCRIPTION
**Pull request by xinzhao3**

### CODE REVIEW

**Comment by bureddy** [comment]
These changes are in a direction to use PUT protocol in cases where GET(READ) protocol is not efficient in GPU Direct RDMA protocols

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4812/ for details (Mellanox internal link).

File: src/ucp/core/ucp_context.c [hunk 1]
```diff
@@ -93,6 +93,11 @@ static ucs_config_field_t ucp_config_table[] = {
    "is zero or negative",
    ucs_offsetof(ucp_config_t, ctx.rndv_thresh_fallback), UCS_CONFIG_TYPE_MEMUNITS},
 
+  {"USE_RNDV_PUT_ZCOPY", "n", "Use put_zcopy in RNDV protocol.\n"
+   "n - Do not use put_zcopy in RNDV protocol, use default get_zcopy"
+   "y - Use put_zcopy in RNDV protocol",
+   ucs_offsetof(ucp_config_t, ctx.use_rndv_put_zcopy), UCS_CONFIG_TYPE_BOOL},
```
Comment (evgeny-leksikov) [inline src/ucp/core/ucp_context.c]: I think context is too global place for protocol switching. Maybe better to make the decision per EP?

File: src/ucp/tag/rndv.c [hunk 2]
```diff
@@ -585,6 +695,97 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_rndv_progress_bcopy_send, (self),
...
+
+    if ((offset == 0) && ((uintptr_t)rndv_req->send.buffer % align) &&
+        (rndv_req->send.length > ucp_mtu )) {
+        length = ucp_mtu - ((uintptr_t)rndv_req->send.buffer % align);
+    } else {
+        length = ucs_min(rndv_req->send.length - offset,
+                         ucp_ep_config(rndv_req->send.ep)->tag.rndv.max_put_zcopy);
+    }
+
+    iov[0].buffer = (void*)rndv_req->send.buffer + offset;
```
Comment (evgeny-leksikov) [inline src/ucp/tag/rndv.c]: Pls use `ucp_dt_iov_copy_uct`

File: src/ucp/tag/rndv.c [hunk 3]
```diff
@@ -585,6 +695,97 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_rndv_progress_bcopy_send, (self),
...
+    iov[0].length = length;
+    iov[0].memh   = rndv_req->send.state.dt.contig.memh;
+    iov[0].count  = 1;
+    iov[0].stride = 0;
+    rndv_req->send.uct_comp.count++;
```
Comment (evgeny-leksikov) [inline src/ucp/tag/rndv.c]: Pls use `ucp_request_send_state_advance`. it will simplify logic below

File: src/ucp/tag/rndv.c [hunk 4]
```diff
@@ -689,6 +890,29 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_rndv_rtr_handler,
...
+        ucp_request_t *rndv_req = ucp_worker_allocate_reply(ep->worker,
+                                                            rndv_rtr_hdr->recv_uuid);
+
+        UCS_PROFILE_CALL(uct_rkey_unpack, rndv_rtr_hdr + 1,
+                         &rndv_req->send.rndv_put.rkey_bundle);
+        rndv_req->send.uct.func       = ucp_proto_progress_rndv_put_zcopy;
+        rndv_req->send.buffer         = sreq->send.buffer;
+        rndv_req->send.length         = sreq->send.length;
+        rndv_req->send.uct_comp.func  = ucp_rndv_put_completion;
+        rndv_req->send.uct_comp.count = 0;
```
Comment (evgeny-leksikov) [inline src/ucp/tag/rndv.c]: Pls use `ucp_request_send_state_reset`

File: src/ucp/tag/rndv.c [hunk 5]
```diff
@@ -67,6 +68,30 @@ size_t ucp_tag_rndv_pack_rkey(ucp_request_t *sreq, ucp_lane_index_t lane,
...
+{
+    ucs_status_t status;
+
+    ucs_assert(UCP_DT_IS_CONTIG(rreq->recv.datatype));
+
+    /* Check if the sender needs to register the send buffer -
+     * is its datatype contiguous and does the receive side need it */
+    if (ucp_ep_rndv_md_flags(ep) & UCT_MD_FLAG_NEED_RKEY) {
+        status = ucp_request_recv_buffer_reg(rreq, ep, lane);
+        ucs_assert_always(status == UCS_OK);
```
Comment (evgeny-leksikov) [inline src/ucp/tag/rndv.c]: Can we propagate the error to user level?

File: src/ucp/tag/rndv.c [hunk 6]
```diff
@@ -585,6 +695,97 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_rndv_progress_bcopy_send, (self),
...
+    if (!(ucp_tag_rndv_is_get_put_op_possible(rndv_req->send.ep, rndv_req->send.lane,
+                                              rndv_req->send.rndv_put.rkey_bundle.rkey))) {
+        /* can't perform put_zcopy - switch to AM rndv */
+        if (rndv_req->send.rndv_put.rkey_bundle.rkey != UCT_INVALID_RKEY) {
+            uct_rkey_release(&rndv_req->send.rndv_put.rkey_bundle);
+        }
+
+        /* switch to AM */
+        if ((UCP_DT_IS_CONTIG(rndv_req->send.datatype)) &&
+            (rndv_req->send.length >= ucp_ep_config(rndv_req->send.ep)->am.zcopy_thresh[0])) {
```
Comment (evgeny-leksikov) [inline src/ucp/tag/rndv.c]: Can `ucp_proto_get_zcopy_threshold` be useful here?

File: src/ucp/tag/rndv.c [hunk 7]
```diff
@@ -426,6 +517,11 @@ UCS_PROFILE_FUNC_VOID(ucp_rndv_matched, (worker, rreq, rndv_rts_hdr),
              * rndv lane */
             ucp_rndv_handle_recv_contig(rndv_req, rreq, rndv_rts_hdr);
         } else {
+            if (ep->worker->context->config.ext.use_rndv_put_zcopy &&
+                ucp_ep_is_rndv_lane_present(ep)) {
```
Comment (alinask) [inline src/ucp/tag/rndv.c]: maybe worth checking here that (rndv_rts_hdr->address != 0)

**Comment by alinask** [comment]
was this tested at scale?
is there a gtest that covers put_zcopy with rndv?

**Comment by xinzhao3** [comment]
@alinask it is not tested with scale yet, will test asap. There is no gtest cover this put_zcopy. Will add one.

**Comment by alinask** [comment]
@xinzhao3 thanks! most of the rndv tests in gtest are in the test_ucp_tag_xfer.cc file.
It has many combinations of datatypes and thresholds to cover many cases.

**Comment by xinzhao3** [comment]
bot:retest

**Comment by xinzhao3** [comment]
@evgeny-leksikov @alinask could you review the pr again? I added test cases to gtest, also test this branch up to 512 processes.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4994/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4995/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4996/ for details (Mellanox internal link).

File: src/ucp/tag/rndv.c [hunk 8]
```diff
@@ -574,6 +686,90 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_rndv_progress_bcopy_send, (self),
...
+    }
+
+    if (!(ucp_tag_rndv_is_get_put_op_possible(rndv_req->send.ep, rndv_req->send.lane,
+                                              rndv_req->send.rndv_put.rkey_bundle.rkey))) {
+        /* can't perform put_zcopy - switch to AM rndv */
+        if (rndv_req->send.rndv_put.rkey_bundle.rkey != UCT_INVALID_RKEY) {
+            uct_rkey_release(&rndv_req->send.rndv_put.rkey_bundle);
+        }
+
+        /* switch to AM */
```
Comment (evgeny-leksikov) [inline src/ucp/tag/rndv.c]: Is it possible to choose the exact protocol by thresholds calculation before stating the request?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5031/ for details (Mellanox internal link).

**Comment by xinzhao3** [comment]
@evgeny-leksikov @alinask could you review again? I changed to use environment variable: get_zcopy, put_zcopy, auto. Currently default is get_zcopy, because this PR is orthogonal to CUDA changes. When CUDA changes is added, I will make auto as default option.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5038/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5039/ for details (Mellanox internal link).

**Comment by xinzhao3** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5049/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@alinask pls take a look

File: src/ucp/core/ucp_context.c [hunk 9]
```diff
@@ -36,6 +36,13 @@ static const char * ucp_device_type_names[] = {
     [UCT_DEVICE_TYPE_SELF] = "loopback",
 };
 
+static const char * ucp_rndv_modes[] = {
+    [UCP_RNDV_MODE_GET_ZCOPY] = "get_zcopy",
+    [UCP_RNDV_MODE_PUT_ZCOPY] = "put_zcopy",
+    [UCP_RNDV_MODE_AUTO]      = "auto",
```
Comment (evgeny-leksikov) [inline src/ucp/core/ucp_context.c]: I do not see where "auto" value is handled.. Below, there are checks in if-statements to exact "put" or "get" values with 'else' cases. So, if user defines this value, behaviour may be erroneous, right? Pls, add a test for this value too.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5089/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5090/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5091/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5092/ for details (Mellanox internal link).

File: src/ucp/tag/rndv.c [hunk 10]
```diff
@@ -66,6 +66,30 @@ size_t ucp_tag_rndv_pack_rkey(ucp_request_t *sreq, ucp_lane_index_t lane,
...
+
+    ucs_assert(UCP_DT_IS_CONTIG(rreq->recv.datatype));
+
+    /* Check if the receiver needs to register the recv buffer -
+     * is its datatype contiguous and does the sender side need it */
+    if (ucp_ep_rndv_md_flags(ep) & UCT_MD_FLAG_NEED_RKEY) {
+        status = ucp_request_recv_buffer_reg(rreq, ep, lane);
+        ucs_assert_always(status == UCS_OK);
+
+        /* if the send buffer was registered, send the rkey */
```
Comment (alinask) [inline src/ucp/tag/rndv.c]: _send_ buffer?

File: src/ucp/tag/rndv.c [hunk 11]
```diff
@@ -574,6 +693,71 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_rndv_progress_bcopy_send, (self),
...
+    }
+
+    rsc_index = ucp_ep_get_rsc_index(rndv_req->send.ep, rndv_req->send.lane);
+    align     = rndv_req->send.ep->worker->ifaces[rsc_index].attr.cap.put.opt_zcopy_align;
+    ucp_mtu   = rndv_req->send.ep->worker->ifaces[rsc_index].attr.cap.put.align_mtu;
+
+    ucs_trace_data("ep: %p try to progress put_zcopy for rndv get. rndv_req: %p. lane: %d",
+                   rndv_req->send.ep, rndv_req, rndv_req->send.lane);
+
+    /* rndv_req is the internal request to perform the get operation */
```
Comment (alinask) [inline src/ucp/tag/rndv.c]: _put_ operation

File: src/ucp/core/ucp_request.h [hunk 12]
```diff
@@ -124,6 +126,13 @@ struct ucp_request {
                 } rndv_get;
 
                 struct {
+                    uint64_t      remote_address; /* address of the sender's data buffer */
+                    uintptr_t     remote_request; /* pointer to the sender's send request */
+                    uct_rkey_bundle_t rkey_bundle;
```
Comment (alinask) [inline src/ucp/core/ucp_request.h]: shouldn't these be _receiver's_ buffer and request?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5192/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5193/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5194/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5195/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5196/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5197/ for details (Mellanox internal link).

**Comment by xinzhao3** [comment]
bot:retest

**Comment by xinzhao3** [comment]
@alinask @evgeny-leksikov I addressed all the comments, could you review the PR again?
There is one failure (malloc_hook_cplusplus.mmmap_ptrs) in Jenkins MellanoxLab, I think it is irrelevant with this PR.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5198/ for details (Mellanox internal link).

File: src/ucp/tag/rndv.c [hunk 13]
```diff
@@ -79,15 +103,16 @@ static size_t ucp_tag_rndv_rts_pack(void *dest, void *arg)
     rndv_rts_hdr->sreq.reqptr      = (uintptr_t)sreq;
     rndv_rts_hdr->sreq.sender_uuid = ep->worker->uuid;
     rndv_rts_hdr->size             = sreq->send.length;
-    if (UCP_DT_IS_CONTIG(sreq->send.datatype)) {
+    if (UCP_DT_IS_CONTIG(sreq->send.datatype) &&
+        (ep->worker->context->config.ext.rndv_mode == UCP_RNDV_MODE_GET_ZCOPY ||
+         ep->worker->context->config.ext.rndv_mode == UCP_RNDV_MODE_AUTO)) {
```
Comment (evgeny-leksikov) [inline src/ucp/tag/rndv.c]: Does current implementation assume `AUTO` == `GET_ZCOPY`? maybe better to handle this on initialization? I mean 
```
if (rndv_mode == UCP_RNDV_MODE_AUTO) {
    rndv_mode = UCP_RNDV_MODE_GET_ZCOPY;
}
```

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5205/ for details (Mellanox internal link).

**Comment by xinzhao3** [comment]
@evgeny-leksikov yeah currently AUTO == GET_ZCOPY. I changed to set it in initialization, could you review again?

**Pull request merged**